test/integration/consul-container/upgrade/basic_test.go

```diff
@@ -0,0 +1,253 @@
+					log_level="TRACE"`,
+				Cmd:     []string{"agent", "-client=0.0.0.0"},
+				Version: *curImage,
+			})
+		require.NoError(t, err)
+	}
+	err = Cluster.AddNodes(Clients)
+	client := Cluster.Nodes[0].GetClient()
+	retry.RunWith(&retry.Timer{Timeout: retryTimeout, Wait: retryFrequency}, t, func(r *retry.R) {
+		leader, err := Cluster.Leader()
+		require.NoError(r, err)
+		require.NotEmpty(r, leader)
+		members, err := client.Agent().Members(false)
+		require.Len(r, members, 5)
+	})
+	err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{Name: "api", Port: 9999})
+	require.NoError(t, err)
+	service, meta, err := client.Catalog().Service("api", "", &api.QueryOptions{})
+	require.NoError(t, err)
+	require.Len(t, service, 1)
+	require.Equal(t, "api", service[0].ServiceName)
+	require.Equal(t, 9999, service[0].ServicePort)
+
+	ch := make(chan struct{})
+	go func() {
+		service, _, err := client.Catalog().Service("api", "", &api.QueryOptions{WaitIndex: meta.LastIndex})
+		require.NoError(t, err)
+		require.Len(t, service, 1)
+		require.Equal(t, "api", service[0].ServiceName)
+		require.Equal(t, 9998, service[0].ServicePort)
+		close(ch)
+	}()
+	err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{Name: "api", Port: 9998})
+	timer := time.NewTimer(1 * time.Second)
+	select {
+	case <-ch:
+	case <-timer.C:
+		t.Fatalf("test timeout")
+	}
+
+}
+
+func TestCurrentServersWithLatestGAClients(t *testing.T) {
+	t.Parallel()
+	numServers := 3
+	Cluster, err := serversCluster(t, numServers, *curImage)
+	require.NoError(t, err)
+	defer Terminate(t, Cluster)
+	numClients := 2
+	Clients := make([]consulNode.Node, numClients)
+	for i := 0; i < numClients; i++ {
+		Clients[i], err = consulNode.NewConsulContainer(context.Background(),
+			consulNode.Config{
+				HCL: `node_name="` + utils.RandName("consul-client") + `"
+					log_level="TRACE"`,
+				Cmd:     []string{"agent", "-client=0.0.0.0"},
+				Version: *curImage,
+			})
+	}
+	client := Cluster.Nodes[0].GetClient()
+	err = Cluster.AddNodes(Clients)
+	retry.RunWith(&retry.Timer{Timeout: retryTimeout, Wait: retryFrequency}, t, func(r *retry.R) {
+		leader, err := Cluster.Leader()
+		require.NoError(r, err)
+		require.NotEmpty(r, leader)
+		members, err := client.Agent().Members(false)
+		require.Len(r, members, 5)
+	})
+	err = client.Agent().ServiceRegister(&api.AgentServiceRegistration{Name: "api", Port: 9999})
+	require.NoError(t, err)
+	service, meta, err := client.Catalog().Service("api", "", &api.QueryOptions{})
+	require.NoError(t, err)
+	require.Len(t, service, 1)
+	require.Equal(t, "api", service[0].ServiceName)
+	require.Equal(t, 9999, service[0].ServicePort)
+
+	ch := make(chan struct{})
+	go func() {
+		service, _, err := client.Catalog().Service("api", "", &api.QueryOptions{WaitIndex: meta.LastIndex})
+		require.NoError(t, err)
+		require.Len(t, service, 1)
+		require.Equal(t, "api", service[0].ServiceName)
+		require.Equal(t, 9998, service[0].ServicePort)
```

Comment:
`t.Fatalf` (or things that wrap it like require) should not be used from goroutines. If you need to do an assertion then change things to be more like:
```
var (
  errCh = make(chan error)
  resultCh = make(chan RESULTTYPE)
)
go func() {
    res, err := STUFF
    if err != nil {
       errCh <- err
    } else {
      resultCh <- res
    }
}()

select {
case err := <-errCh:
     require.NoError(t, err)
case res := <-resultCh:
    < all your require assertions>
case <-timer.C:
    die
    }
    ```
}